### PR TITLE
Faster axis autorange + remove calcIfAutorange edit type

### PR DESCRIPTION
**Pull request by etpinard**
resolves https://github.com/plotly/plotly.js/issues/2546 and brings cartesian axis autorange relayout calls (e.g. on-plot-area double-click or the "autoscale" mode bar button) to perform to a level comparable to axis range relayout calls (e.g zoom in/out).

In brief, cartesian `axis.autorange` is now declared as `editType: 'axrange'` and edit type 'calcIfAutorange' is gone.

As mentioned in https://github.com/plotly/plotly.js/issues/2546#issuecomment-405388288, adapting the update pathway ended up being a little trickier than I'd imagine (especially for annotations, see https://github.com/plotly/plotly.js/commit/a8fb653ddca6dd00407b4362131ba2ceb3e6d511). But I'm pretty please with the results: a lot of old delicate "find-autoranged-axes" logic in restyle, relayout and react is gone (see https://github.com/plotly/plotly.js/commit/138a84f613c7c6185c6c1d1d3978fda4754f76e5)

However, there's one drawback that is important to mention. While this PR speeds up axis autorange relayouts, it does _slow down_:
- first render for graphs with set axis ranges (as `Axes.expand` is no longer skipped), and 
- updates of old 'calcIfAutorange' attributes (that are now declared as `editType: 'calc'`) on graphs with set ranges. The most significant (I think) is scatter `marker.size` where updating its values will now **always** lead to a calc edit pathway. 

It would be nice to open a new edit pathway with a targeted  `calc` step where only the relevant `_module.calc` / `_module.setPositions` or component `calcAutorange` would be called. But for the time being, I much prefer having fast double-click and mode bar interactions than fast `marker.size` restyle calls. I hope you agree. 

This PR should be part of the next minor release 1.40.0, not a patch release in the 1.39.x series.

cc @alexcjohnson

### CODE REVIEW

File: test/jasmine/tests/finance_test.js [hunk 1]
```diff
@@ -931,6 +931,58 @@ describe('finance charts updates:', function() {
...
+                tickLen: 0.3,
+                pathd: 'M-20.25,134.55H0M0,81V193.5M20.25,85.5H0'
+            });
+            return Plotly.restyle(gd, 'tickwidth', 0);
+        })
+        .then(function() {
+            _assert('set rng / no tickwidth', {
+                tickLen: 0,
+                pathd: 'M0,134.55H0M0,81V193.5M0,85.5H0'
+            });
```
Comment (etpinard) [inline test/jasmine/tests/finance_test.js]: This asserting here fails on `master`, so count this as a bug fix.

File: src/plot_api/plot_api.js [hunk 2]
```diff
@@ -1708,6 +1708,7 @@ function addAxRangeSequence(seq, rangesAltered) {
         subroutines.doTicksRelayout;
 
     seq.push(
+        subroutines.doAutoRangeAndConstraints,
```
Comment (etpinard) [inline src/plot_api/plot_api.js]: Adding `doAutoRangeAndConstraints` to the 'axrange' update sequence does work against some of the work done in https://github.com/plotly/plotly.js/pull/2628, but only by 1ms in the most extremes case (a 50x50 splom) when there are no axis constraints.

File: test/jasmine/tests/axes_test.js [hunk 3]
```diff
@@ -1858,24 +1858,21 @@ describe('Test axes', function() {
             expect(ax._max).toEqual([{val: 6, pad: 10, extrapad: true}]);
         });
 
-        it('should return early if no data is given', function() {
+        it('should fail if no data is given', function() {
```
Comment (alexcjohnson) [inline test/jasmine/tests/axes_test.js]: you're confident we always send an array to `expand`? I guess in principle if we don't have a required array the trace will be set `visible: false`...

File: src/traces/scatter/calc.js [hunk 4]
```diff
@@ -96,8 +96,9 @@ function calcAxisExpansion(gd, trace, xa, ya, x, y, ppad) {
         yOptions.padded = false;
     }
 
-    Axes.expand(xa, x, xOptions);
-    Axes.expand(ya, y, yOptions);
+    // N.B. asymmetric splom traces call this with undefined xa or ya
```
Comment (alexcjohnson) [inline src/traces/scatter/calc.js]: not `undefined`, that would error here as well as above... it's [`{}`](https://github.com/plotly/plotly.js/blob/24b291285a15650a76ae6c78e2762a8e88f2027d/src/traces/splom/index.js#L66-L67)

File: src/components/annotations/attributes.js [hunk 5]
```diff
@@ -19,7 +19,7 @@ module.exports = templatedArray('annotation', {
         valType: 'boolean',
         role: 'info',
         dflt: true,
-        editType: 'calcIfAutorange+arraydraw',
+        editType: 'calc+arraydraw',
```
Comment (alexcjohnson) [inline src/components/annotations/attributes.js]: Not for now of course, but annotations (showing, hiding, dragging in GUI) are a great argument in favor of each trace and item tracking its own autorange contributions, so it doesn't take a full recalc to move one annotation.

File: test/jasmine/tests/annotations_test.js [hunk 6]
```diff
@@ -737,6 +737,43 @@ describe('annotations autorange', function() {
...
+        })
+        .then(function() {
+            _assert('auto rng / big tx', [-0.22, 3.57], [0.84, 3.365]);
+            return Plotly.relayout(gd, 'annotations[0].text', 'a');
+        })
+        .then(function() {
+            _assert('auto rng / small tx', [-0.18, 3.035], [0.84, 3.365]);
+        })
+        .catch(failTest)
+        .then(done);
```
Comment (alexcjohnson) [inline test/jasmine/tests/annotations_test.js]: Beautiful test!

File: src/components/annotations/draw.js [hunk 7]
```diff
@@ -250,11 +250,6 @@ function drawRaw(gd, options, index, subplotId, xa, ya) {
         var outerWidth = Math.round(annWidth + 2 * borderfull);
         var outerHeight = Math.round(annHeight + 2 * borderfull);
 
-
-        // save size in the annotation object for use by autoscale
-        options._w = annWidth;
-        options._h = annHeight;
```
Comment (alexcjohnson) [inline src/components/annotations/draw.js]: Good catch, great to see all the pruning you manage to do in this PR!

**Comment by alexcjohnson** [comment]
@etpinard this is great, I have no qualms about this tradeoff and anyway we have a route to improve it even more later on with per-trace updates. A couple of minor comments but nothing blocking. 💃

File: test/jasmine/tests/axes_test.js (hunk 3 above)
Comment (etpinard) [reply test/jasmine/tests/axes_test.js]: Yes, `expand` is only called from trace module `calc` or `setPosition` or component `calcAutorange` which are skipped for `visible: false` traces / component items.

I thought removing that `if(!data) return;` was a nice way to `expand` is only called when needed.

File: src/components/annotations/attributes.js (hunk 5 above)
Comment (etpinard) [reply src/components/annotations/attributes.js]: > so it doesn't take a full recalc to move one annotation.

Absolutely! I'll open an issue about this once this PR is merged.

File: src/traces/scatter/calc.js (hunk 4 above)
Comment (etpinard) [reply src/traces/scatter/calc.js]: good catch -> 026cd53

**Comment by etpinard** [comment]
Ok let's start merging things for 1.40.x

**Pull request merged**